### PR TITLE
[INFINITY-2600] Add kafka setup helper

### PR DESCRIPTION
**Pull request by benclarkwood**
This PR sets the stage for adding authz settings and the TLS authn settings.

The existing toggling was getting a bit hairier than Mustache can accomodate. This PR:
1. Creates a new go utility that calculates the listeners and advertised listeners.
2. Drops setting the broker "port" as that is done with listeners.
3. The go utility also calculates the inter broker protocol.
4. A bunch of tests for the go utility.

With this PR the following are valid combinations of settings:

Kerberos on
Kerberos and TLS transport encryption on
TLS transport encryption on
TLS transport encryption on and Plaintext allowed

### CODE REVIEW

File: frameworks/kafka/setup-helper/setup_helper_test.go [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+	os.Setenv(brokerPort, "1000")
+
+	asrt.Equal("PLAINTEXT://127.0.0.1:1000", getListener("PLAINTEXT", brokerPort))
+	os.Clearenv()
+}
+
+func TestGetListenerTLS(t *testing.T) {
+	asrt := assert.New(t)
+
+	os.Setenv(ipEnvvar, "127.0.0.1")
```
Comment (elezar) [inline frameworks/kafka/setup-helper/setup_helper_test.go]: Would it make sense to use a different IP as well?

File: frameworks/kafka/setup-helper/setup_helper.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+		} else {
+			protocol = "SASL_PLAINTEXT"
+		}
+	} else if tlsEncryptionEnabled {
+		protocol = "SSL"
+	} else {
+		protocol = "PLAINTEXT"
+	}
+
+	err = writeToWorkingDirectory(property, fmt.Sprintf("%s=%s", property, protocol))
```
Comment (elezar) [inline frameworks/kafka/setup-helper/setup_helper.go]: For the listener configs, we use a different naming convention for the file.

File: frameworks/kafka/src/main/dist/server.properties.mustache [hunk 3]
```diff
@@ -27,23 +27,17 @@ broker.id={{POD_INSTANCE_INDEX}}
...
-{{/KAFKA_ENABLE_TLS}}
-{{/SECURITY_KERBEROS_ENABLED}}
-
+#
+# Hostname and port the broker will advertise to producers and consumers. If not set,
+# it uses the value for "listeners" if configured.  Otherwise, it will use the value
+# returned from java.net.InetAddress.getCanonicalHostName().
+#
+# advertised.listeners=PLAINTEXT://your.host.name:9092
+{{ADVERTISED_LISTENERS}}
```
Comment (elezar) [inline frameworks/kafka/src/main/dist/server.properties.mustache]: Oh this is SEXY!

File: frameworks/kafka/src/main/dist/svc.yml [hunk 4]
```diff
@@ -70,15 +71,31 @@ pods:
...
+          export ADVERTISED_LISTENERS=`cat advertised-listeners-config`
+          export LISTENERS=`cat listeners-config`
+          export SECURITY_INTER_BROKER_PROTOCOL=`cat security.inter.broker.protocol`
+
+          ./bootstrap -resolve=false
+
+          # NOTE: We add some custom statsd libraries for statsd metrics as well
+          # as a custom zookeeper library to support our own ZK running
+          # kerberized. The custom zk library does not do DNS reverse resolution
+          # of the ZK hostnames.
```
Comment (elezar) [inline frameworks/kafka/src/main/dist/svc.yml]: Is this comment not duplicated now?

File: frameworks/kafka/setup-helper/setup_helper.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+func setListeners() error {
+	var listeners []string
+	var advertisedListeners []string
+
+	kerberosEnabled, tlsEncryptionEnabled, allowPlainText, err := parseToggles()
+	if err != nil {
+		return err
+	}
+
+	if kerberosEnabled { // Kerberos enabled
```
[the remark this answers is not in the document]
Comment (benclarkwood) [reply frameworks/kafka/setup-helper/setup_helper.go]: Hm. 

You could switch the logic to be:
"What listeners do I need to have?"

Then feed that list to a function that spits out the listeners.

File: frameworks/kafka/setup-helper/setup_helper.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+	brokerPort            = "KAFKA_BROKER_PORT"
+	brokerPortTLS         = "KAFKA_BROKER_PORT_TLS"
+	taskNameEnvvar        = "TASK_NAME"
+	frameworkHostEnvvar   = "FRAMEWORK_HOST"
+	ipEnvvar              = "MESOS_CONTAINER_IP"
+	kerberosPrimaryEnvvar = "SECURITY_KERBEROS_PRIMARY"
+)
+
+func main() {
+	log.Printf("Starting setup-helper...")
```
[the remark this answers is not in the document]
Comment (benclarkwood) [reply frameworks/kafka/setup-helper/setup_helper.go]: Dependencies tho!

File: frameworks/kafka/setup-helper/setup_helper.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	log.Printf("Calculating security and listener settings...")
+	err := calculateSettings()
+	if err != nil {
+		log.Fatalf("Failed to calculate security and listener settings: %s", err.Error())
+	}
+	log.Printf("Calculated security and listener settings.")
+	log.Printf("setup-helper complete.")
+}
+
+func getBooleanEnvvar(envvar string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (benclarkwood) [reply frameworks/kafka/setup-helper/setup_helper.go]: Fixed!

**Comment by benclarkwood** [comment]
Closed so CI doesn't trigger

**Comment by benclarkwood** [comment]
@elezar: I chose to not separate writing from calculating at the time, and rewriting it now is probably non-zero time.

I also didn't address the probably simpler suggestion you had for calculating listeners.

I'll open a JIRA for that.

**Comment by benclarkwood** [comment]
https://jira.mesosphere.com/browse/INFINITY-2768

**Pull request merged**